Documentation/design-docs/variabletracking.md

```diff
@@ -0,0 +1,314 @@
+Reporting Variable Location
+===================
+Table of Contents
+-----------------
+
+[Debug Info](#debug-info)
+
+[Context](#context)
+
+[Debug Code vs Optimized Code](#debug-code-vs-optimized-code)
+
+[siScope / psiScope Structures](#siscope-psiscope-structure)
+
+[Generating siScope / psiScope info](#siscope-psiscope)
+
+[Variable Live Range Structure](#variable-live-range-structure)
+
+[Generating Variable Live Range](#variable-live-range)
+
+[Turning On Debug Info](#debug-indo-flags)
+
+[Dumps and Debugging Support](#dumps-and-debugging-support)
+
+[Future Extensions and Enhancements](#future-extensions-and-enhancements)
+
+Debug Info
+--------
+
+The debugger expects to receive an array of `VarResultInfo` which indicates:
+
+-   IL variable number
+
+-   Location
+
+-   First native offset this location is valid
+
+-   First native offset this location is no longer valid
+
+There could be more than one per variable.
+There should be no overlap in terms of starting/ending native offsets for the same variable.
+Given a native offset, no two variables should share the same location.
+
+Context  
+--------
+
+We generate variables debug info while we generate the assembly intructions of the code. This is happening in `genCodeForBBList()` and `genFnProlog()`.
+
+If we are jitting release code, information about variable liveness will be included in `GenTree`s and `BasicBlock` when `genCodeForBBList` is reached.
+For each When `BasicBlock`:
+
+-   `bbLiveIn`: alive at the beginning
+
+-   `bbLiveOut`: alive at the end
+
+Also each `GenTree` has a mask `gtFlags` indicating will include if a variable:
+
+-   is being born
+
+-   is becoming death
+
+-   is being spilled
+
+-   has been spilled
+
+When generating each instruction, this flags are used to know if we need to start/end tracking a variable or update its location (which means in practice "end and start").
+The blocks set of live variables are also used to start/end variables tracking information. 
+
+Once the code is done we know the native offset of every assembly instruction, which is required for the debug info.
+While we are generating code we don't work with native offsets but `emitLocation`s that the `emitter` can generate, and once the code is done we can get the native offset corresponding to each of them.
+
+Debug Code vs Optimized Code
+--------
+
+Variables on debug code:
+-   are alive during the whole method
+
+-   live in a fixed location on the stack the whole method
+
+-   `bbLiveIn` and `bbLiveOut` sets are empty
+
+-   there is no flag for spill/unspill variables
+
+Variables on optimized code:
+
+-   could change their location during code execution
+
+-   could change their liveness during code execution
+
+siScope / psiScope Structures
+--------
+
+#### siScope
+
+This struct is used to represent the ranges where a variable is alive during code generation.
+
+It has two field to denote the native offset where it is valid:
```

Comment:
"It" is a little ambiguous. Maybe better to be explicit: "The struct has two fields to denote the native offset where a variable is valid"